### PR TITLE
Ccde 549 metadata joins

### PR DESCRIPTION
**Pull request by mratford**

### CODE REVIEW

File: tests/test_metadata.py [hunk 1]
```diff
@@ -570,6 +570,112 @@ def test_inferred_input_fails(fake_input):
...
+
+
+@pytest.mark.parametrize(
+    "m1,m2",
+    [
+        (merge_meta_test, merge_meta_diff_col_type),
+        (merge_meta_test, merge_meta_diff_partitions),
+    ],
+)
+def test_merge_error_raised(m1, m2):
```
Comment (s-ducks) [inline tests/test_metadata.py:613]: Can you add another test that verifies that using `mismatch=error` doesn't cause an error when it isn't supposed to

**Review by s-ducks — CHANGES_REQUESTED** [review]

File: mojap_metadata/metadata/metadata.py [hunk 2]
```diff
@@ -221,6 +222,55 @@ def from_infer(cls, inp, **kwargs) -> object:
         else:
             raise TypeError(f"input type not recognised: {type(inp)}")
 
+    @classmethod
+    def merge(cls, old, new, mismatch="priority", data={}) -> object:
```
[the remark this answers is not in the document]
Comment (mratford) [reply mojap_metadata/metadata/metadata.py]: https://github.com/moj-analytical-services/mojap-metadata/blob/d29361c105e7006a7b7deef4398eb503e67ea146/mojap_metadata/metadata/metadata.py#L224-L286

File: mojap_metadata/metadata/metadata.py [hunk 3]
```diff
@@ -221,6 +222,55 @@ def from_infer(cls, inp, **kwargs) -> object:
...
+            new: Metadata object
+            mismatch: Determines behaviour when there is a mismatch of column
+                metadata, "priority" prioritises the details from the
+                new Metadata object, "error" throws an error
+            data: Metadata values for the merged Metadata object, overriding
+                existing values
+        returns:
+            Metadata object
+        """
+        m = deepcopy(old)
```
[the remark this answers is not in the document]
Comment (mratford) [reply mojap_metadata/metadata/metadata.py]: https://github.com/moj-analytical-services/mojap-metadata/blob/d29361c105e7006a7b7deef4398eb503e67ea146/mojap_metadata/metadata/metadata.py#L219-L220

File: mojap_metadata/metadata/metadata.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mratford) [reply mojap_metadata/metadata/metadata.py]: https://github.com/moj-analytical-services/mojap-metadata/blob/d29361c105e7006a7b7deef4398eb503e67ea146/mojap_metadata/metadata/metadata.py#L224-L253

File: mojap_metadata/metadata/metadata.py [hunk 4]
```diff
@@ -221,6 +222,55 @@ def from_infer(cls, inp, **kwargs) -> object:
...
+        for k in data:
+            m._data[k] = data[k]
+
+        if mismatch == "error":
+            for k in [
+                x
+                for x in m._data
+                if x in new._data and x != "columns" and x not in data
+            ]:
```
[the remark this answers is not in the document]
Comment (mratford) [reply mojap_metadata/metadata/metadata.py]: https://github.com/moj-analytical-services/mojap-metadata/blob/d29361c105e7006a7b7deef4398eb503e67ea146/mojap_metadata/metadata/metadata.py#L255-L262

File: mojap_metadata/metadata/metadata.py [hunk 5]
```diff
@@ -413,6 +464,14 @@ def _validate_list_attribute(self, attribute: str, columns: list) -> None:
         if len(columns) != len(set(columns)):
             raise ValueError(f"'All elements of '{attribute}' must be unique")
 
+    def _validate_unique_column_names(self) -> None:
```
[the remark this answers is not in the document]
Comment (mratford) [reply mojap_metadata/metadata/metadata.py]: https://github.com/moj-analytical-services/mojap-metadata/blob/d29361c105e7006a7b7deef4398eb503e67ea146/mojap_metadata/metadata/metadata.py#L255-L262

File: tests/test_metadata.py (hunk 1 above)
Comment (mratford) [reply tests/test_metadata.py:613]: https://github.com/moj-analytical-services/mojap-metadata/blob/d29361c105e7006a7b7deef4398eb503e67ea146/tests/test_metadata.py#L618-L656

**Review by s-ducks — APPROVED** [review]

**Pull request merged**